src/Streamly/Coreutils/Ln.hs

```diff
@@ -0,0 +1,52 @@
+-- |
+-- Module      : Streamly.Coreutils.Ln
+-- Copyright   : (c) 2022 Composewell Technologies
+-- License     : Apache-2.0
+-- Maintainer  : streamly@composewell.com
+-- Stability   : experimental
+-- Portability : GHC
+--
+-- create a link to TARGET with the given name
+
+module Streamly.Coreutils.Ln
+    (
+      Ln
+    , ln
+    , force
+    , symbolic
+    )
+where
+
+import Control.Monad (when)
+import Streamly.Coreutils.Common (Switch(..))
+import Streamly.Coreutils.FileTest (test, isExisting)
+
+import qualified System.Posix.Files as Posix
+
+data Ln = Ln
+    { lnForce :: Switch
+    , lnSymbolic :: Switch
+    }
+
+defaultConfig :: Ln
+defaultConfig = Ln Off Off
+
+force :: Switch -> Ln -> Ln
+force opt cfg = cfg {lnForce = opt}
+
+symbolic :: Switch -> Ln -> Ln
+symbolic opt cfg = cfg {lnSymbolic = opt}
+
+ln :: (Ln -> Ln) -> FilePath -> FilePath -> IO ()
+ln f src tgt = do
+    let opt = f defaultConfig
+    found <- test tgt isExisting
+    when (lnForce opt == Off) $ do
+        when found $ error msg
+        case lnSymbolic opt of
```

Comment:
This should not be inside the when statement, otherwise we are not doing anything at all if the force option `On`. This should be done unconditionally. If the file exists we will bail out already due to the `error`.